src/interfacers/EmonHubBleInterfacer.py

```diff
@@ -0,0 +1,168 @@
+        device_addr = '00:0b:57:64:8c:a2'
+    [[[runtimesettings]]]
+        pubchannels = ToEmonCMS,
+	read_interval = 20
+
+[nodes]
+
+[[1]]
+    nodename = Sensornode
+
+    [[[rx]]]
+        names = temp,humidity,battery
+        scales = 0.01,0.01,1
+        units = C,%,%
+
+"""
+
+class EmonHubBleInterfacer(EmonHubInterfacer):
+
+    def __init__(self, name, device_addr=''):
+        """Initialize interfacer
+
+        device_addr (string): BLE MAC address to connect to
+
+        """
+
+        # Initialization
+        super(EmonHubBleInterfacer, self).__init__(name)
+
+        self._private_settings = {
+            'read_interval': 60
+        }
+
+        self._addr = device_addr
+        self._last_read_time = 0
+        self._bat_readings = []
+
+	self._connect()
+
+    def close(self):
+        """Close serial port"""
+
+        # Close serial port
+        if self._ble is not None:
+            self._log.debug("Closing Bluetooth connection")
+            self._ble.disconnect()
+
+        return
+
+    def read(self):
+        """Read data from bluetooth sensor
+
+        """
+
+        # Don't read before the configured interval
+        interval = int(self._private_settings['read_interval'])
+        if time.time() - self._last_read_time < interval:
+            return
+
+        self._last_read_time = time.time()
+
+        # Check connection, connect if we didn't connect during init
+        if not self._ble:
+            self._connect()
+
+        if not self._ble:
+            return False
+
+        temp = self._get_temperature()
+        rh = self._get_humidity()
+        bat = self._get_bat_level()
+
+        data = '{}, {}, {}'.format(temp, rh, bat )
+
+        # Create a Payload object
+        c = Cargo.new_cargo(rawdata=data)
+        c.realdata = (temp, rh, bat)
+
+        if int(self._settings['nodeoffset']):
+            c.nodeid = int(self._settings['nodeoffset'])
+        else:
+            c.nodeid = 1
+
+        return c
+
+    def set(self, **kwargs):
+
+        for key, setting in self._private_settings.iteritems():
+            # Decide which setting value to use
+            if key in kwargs.keys():
+                setting = kwargs[key]
+            else:
+                setting = self._private_settings[key]
+
+            # Ignore unchanged
+            if key in self._settings and self._settings[key] == setting:
+                continue
+            elif key == 'read_interval':
+                setting = float(setting)
+            else:
+                self._log.warning("'%s' is not valid for %s: %s" % (str(setting), self.name, key))
+                continue
+
+            self._log.debug('Setting {}: {}'.format(key, setting))
+            self._private_settings[key] = setting
+
+        # include kwargs from parent
+        super(EmonHubBleInterfacer, self).set(**kwargs)
+
+    def _get_temperature(self):
+        val = self._temperature.read()
+        (val,) = struct.unpack('h', val)
```

Comment:
emonhub knows how to decode data using its "datacodes" mechanism. Could we leverage the existing code instead of hard-coding the data types in these methods?

Reply:
possibly, I will leave that modification to someone who might use this interfacer as I done have the ability to test this myself.